chemprop/uncertainty/calibrator.py

```diff
@@ -79,57 +80,220 @@ def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
+    More detailes can be found in [1]_.
+
+    Parameters
+    ----------
+    alpha: float
+        Error rate, :math:`\alpha \in [0, 1]`
+
+    References
+    ----------
+    .. [1] Angelopoulos, A.N.; Bates, S.; "A Gentle Introduction to Conformal Prediction and Distribution-Free
+        Uncertainty Quantification." arXiv Preprint 2021, https://arxiv.org/abs/2107.07511
+    """
+
+    def __init__(self, alpha: float):
+        super().__init__()
+        self.alpha = torch.tensor(alpha)
+        self.qhats = torch.tensor([])
+
+    @staticmethod
+    def nonconformity_scores(preds: Tensor):
+        r"""Compute nonconformity score as the negative of the softmax output for the true class.
+
+        .. math::
+            s_i = -\hat{f}(X_i)_{Y_i}
+        """
+        return -preds
+
     def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+        scores = self.nonconformity_scores(uncs)
+        for j in range(uncs.shape[1]):
+            mask_j = mask[:, j]
+            targets_j = targets[:, j][mask_j]
+            scores_j = scores[:, j][mask_j]
+
+            scores_j = torch.gather(scores_j, 1, targets_j.unsqueeze(1)).squeeze(1)
+            num_data = targets_j.shape[0]
+            q_level = (torch.ceil((num_data + 1) * (1 - self.alpha)) / num_data).item()
+            qhat = torch.quantile(scores_j, q_level, interpolation="higher")
+            self.qhats = torch.cat((self.qhats, torch.tensor([qhat])))
 
     def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        calibrated_preds = torch.zeros_like(uncs, dtype=torch.int)
+        scores = self.nonconformity_scores(uncs)
+
+        for j, qhat in enumerate(self.qhats):
+            calibrated_preds[:, j] = (scores[:, j] <= qhat).int()
+
+        return preds, calibrated_preds
 
 
 @UncertaintyCalibratorRegistry.register("conformal-adaptive")
-class ConformalAdaptiveMulticlassCalibrator(UncertaintyCalibrator):
-    def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+class ConformalAdaptiveMulticlassCalibrator(ConformalMulticlassCalibrator):
+    @staticmethod
+    def nonconformity_scores(preds):
+        r"""Compute nonconformity score by greedily including classes in the classification set until it reach the true label.
 
-    def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        .. math::
+            s(x, y) = \sum_{j=1}^{k} \hat{f}(x)_{\pi_j(x)}, \text{ where } y = \pi_k(x)
+
+        where :math:`\pi_k(x)` is the permutation of :math:`\{1,...,K\}` that sorts :math:`\hat{f}(X_{test})` from most likely to least likely.
+        """
+
+        sort_inds = torch.argsort(-preds, dim=2)
+        sorted_preds = torch.gather(preds, 2, sort_inds)
+        sorted_scores = sorted_preds.cumsum(dim=2)
+        unsort_inds = torch.argsort(sort_inds, dim=2)
```

Comment:
FWIW, I prefer `index` to keep in line with the pytorch argspec

```suggestion
        sorted_index = torch.argsort(-preds, dim=2)
        sorted_preds = torch.gather(preds, dim=2, index=sorted_index)
        sorted_scores = sorted_preds.cumsum(dim=2)
```